server/controllers/users.js

```diff
@@ -92,25 +96,55 @@ export const updateToken = async (req, res) => {
 };
 
 /**
- * Verify the 2FA code the user has entered when
- * logging in and send back another JWT.
+ * Verify the 2FA code or recovery code the user has entered when logging in and send back another JWT.
  */
 export const twoFactorAuthAndUpdateToken = async (req, res, next) => {
-  const { twoFactorAuthenticatorCode } = req.body;
+  const { twoFactorAuthenticatorCode, twoFactorAuthenticationRecoveryCode } = req.body;
 
   const userId = Number(req.jwtPayload.sub);
+
+  // Both 2FA and recovery codes rate limited to 10 tries per hour
+  const rateLimit = new RateLimit(`user_2FA_endpoint_${userId}`, 10, ONE_HOUR_IN_SECONDS);
+  const fail = async exception => {
+    await rateLimit.registerCall();
+    next(exception);
+  };
+
+  if (await rateLimit.hasReachedLimit()) {
+    return next(new Unauthorized('Too many attempts. Please try again in an hour'));
```

Comment:
Seems more appropriate

```suggestion
    return next(new RateLimitExceeded('Too many attempts. Please try again in an hour'));
```